source/extensions/quic_listeners/quiche/envoy_quic_packet_writer.cc

```diff
@@ -1,43 +1,71 @@
 #include "extensions/quic_listeners/quiche/envoy_quic_packet_writer.h"
 
-#include "common/buffer/buffer_impl.h"
-#include "common/network/utility.h"
-
 #include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
 
 namespace Envoy {
 namespace Quic {
-EnvoyQuicPacketWriter::EnvoyQuicPacketWriter(Network::Socket& socket)
-    : write_blocked_(false), socket_(socket) {}
+
+namespace {
+
+quic::WriteResult convertToQuicWriteResult(Api::IoCallUint64Result& result) {
+  if (result.ok()) {
+    return {quic::WRITE_STATUS_OK, static_cast<int>(result.rc_)};
+  }
+  quic::WriteStatus status = result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again
+                                 ? quic::WRITE_STATUS_BLOCKED
+                                 : quic::WRITE_STATUS_ERROR;
+  return {status, static_cast<int>(result.err_->getErrorCode())};
+}
+
+} // namespace
+
+EnvoyQuicPacketWriter::EnvoyQuicPacketWriter(Network::UdpPacketWriterPtr envoy_udp_packet_writer)
+    : envoy_udp_packet_writer_(std::move(envoy_udp_packet_writer)) {}
 
 quic::WriteResult EnvoyQuicPacketWriter::WritePacket(const char* buffer, size_t buf_len,
                                                      const quic::QuicIpAddress& self_ip,
                                                      const quic::QuicSocketAddress& peer_address,
                                                      quic::PerPacketOptions* options) {
   ASSERT(options == nullptr, "Per packet option is not supported yet.");
-  ASSERT(!write_blocked_, "Cannot write while IO handle is blocked.");
 
-  Buffer::RawSlice slice;
-  slice.mem_ = const_cast<char*>(buffer);
-  slice.len_ = buf_len;
+  Buffer::InstancePtr buf;
+  buf = std::make_unique<Buffer::OwnedImpl>(buffer, buf_len);
```

Comment:
This is introducing an extra copy, use  OwnedImpl::addBufferFragment() instead.

Reply:
Using OwnedImpl::add() here to add the char* buffer here.
```
Buffer::OwnedImpl buf;
buf.add(buffer, buf_len);
```

Reply:
This is still copying data from |buffer| IIUC.

Reply:
Thanks for pointing this out Dan! I have updated this to use `addBufferFragment()` now.